### ISSUE TITLE
Not reporting error when none of the overload variants match

### ISSUE THREAD

**Issue by povilasb** [microsoft/pyright#4477]
**Describe the bug**

1. Say a class has a method with multiple overloads.
2. When the method is called with arguments that do not match any of the overloads
2.1. an error is correctly reported by **pyright** up to `1.1.254`,
2.1. no error with versions `>=1.1.255`.
3. **mypy 0.991** reports an error.

**To Reproduce**

1. Get the latest **pyright** (`1.1.289`)
2. Run it on a provided code snippet.
3. The only errors you should see are `Overloaded implementation is not consistent with signature of overload`.
4. Get **pyright** `1.1.254`.
5. Run the checks again and you should see:
```sh
➜  /tmp pyright typing_issues.py
No configuration file found.
No pyproject.toml file found.
stubPath /private/tmp/typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
pyright 1.1.254
/private/tmp/typing_issues.py
  /private/tmp/typing_issues.py:32:57 - error: Could not bind method "for_each" because "Flow[set[str]]" is not assignable to parameter "self"
    TypeVar "OutT@Flow" is invariant
      "set[str]" is incompatible with "list[Out2T@for_each]" (reportGeneralTypeIssues)
  /private/tmp/typing_issues.py:32:57 - error: Could not bind method "for_each" because "Flow[set[str]]" is not assignable to parameter "self"
    TypeVar "OutT@Flow" is invariant
```
6. Optionally you can run **mypy** `0.991` which produces an error too:
```sh
➜  /tmp mypy typing_issues.py
typing_issues.py:32: error: Invalid self argument "Flow[Set[str]]" to attribute function "for_each" with type "Callable[[Flow[List[Out2T]], Callable[[Out2T], NewOutT]], Flow[List[NewOutT]]]"  [misc]
```

**Expected Behavior**

Same as **mypy** and **pyright** pre `1.1.255` version. 

**Code**

```py
from typing import TypeVar, Generic, Callable, overload


OutT = TypeVar("OutT")
Out2T = TypeVar("Out2T")
NewOutT = TypeVar("NewOutT")


class Flow(Generic[OutT]):
    def __init__(self, initial_value: OutT):
        ...

    def map(self: "Flow[OutT]", func: Callable[[OutT], NewOutT]) -> "Flow[NewOutT]":
        ...

    @overload
    def for_each(
        self: "Flow[list[Out2T]]", func: Callable[[Out2T], NewOutT]
    ) -> "Flow[list[NewOutT]]":
        ...

    @overload
    def for_each(
        self: "Flow[tuple[Out2T]]", func: Callable[[Out2T], NewOutT]
    ) -> "Flow[tuple[NewOutT]]":
        ...

    def for_each(self, func):
        ...


flow = Flow("1.2.3.4").map(lambda v: set(v.split("."))).for_each(lambda v: int(v))
```

* A computational flow where `.for_each()` should only be allowed to operate when the previous flow result is either a `tuple` or a `list`.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by povilasb** [comment]
Thank you.

**Comment by erictraut** [comment]
This is included in pyright 1.1.290, which I just published. It will also be included in a future release of pylance.
